### PR TITLE
PR for Carlos

### PR DESCRIPTION
**Pull request by twicki**

### CODE REVIEW

File: src/dawn/IIR/FieldAccessMetadata.cpp [hunk 1]
```diff
@@ -17,6 +17,25 @@
 namespace dawn {
 namespace iir {
 
+std::string toString(FieldAccessType type) {
+  if(type == FieldAccessType::FAT_GlobalVariable) {
```
Comment (twicki) [inline src/dawn/IIR/FieldAccessMetadata.cpp]:
```suggestion
    switch(type){
  case FieldAccessType::FAT_GlobalVariable:
      return "GlobalVariable";
  case FieldAccessType::FAT_Literal:
      return "Literal";
      ...
  default:
      return "Unknown";
  }
```

File: src/dawn/IIR/StencilInstantiation.cpp [hunk 2]
```diff
@@ -47,7 +47,9 @@ namespace iir {
 //===------------------------------------------------------------------------------------------===//
 
 StencilInstantiation::StencilInstantiation(dawn::OptimizerContext* context)
-    : context_(context), IIR_(make_unique<IIR>()), SIR_(context->getSIR()) {}
+    : context_(context), metadata_(*(context->getSIR()->GlobalVariableMap)),
```
Comment (twicki) [inline src/dawn/IIR/StencilInstantiation.cpp:51]: why did we duplicate the `GlobalVariableMap` to `StencilMetaInformation` and `IIR`? since this is first class citizen information I'd suggest only having it in the `IIR`. Thoughts?

File: src/dawn/IIR/StencilInstantiation.h [hunk 3]
```diff
@@ -227,55 +157,22 @@ class StencilInstantiation : NonCopyable {
...
-  /// @brief it finalizes the stencil function instantation. The stencil function instantatiation is
-  /// moved from candidate to the final storage of stencil instantiations. And maps storing
-  /// stencil functions of the stencil instantiation are updated
-  void finalizeStencilFunctionSetup(std::shared_ptr<StencilFunctionInstantiation> stencilFun);
-
-  const std::vector<std::shared_ptr<sir::StencilFunction>>& getStencilFunctions() {
-    return SIR_->StencilFunctions;
-  }
+  /// If `curStencilFunctionInstantiation` is not NULL, the stencil function is treated as a
+  /// nested
```
Comment (twicki) [inline src/dawn/IIR/StencilInstantiation.h]: indentation

File: src/dawn/IIR/StencilMetaInformation.cpp [hunk 4]
```diff
@@ -118,35 +101,232 @@ bool StencilMetaInformation::isAccessType(FieldAccessType fType, const std::stri
...
-           !isAccessType(FieldAccessType::FAT_InterStencilTemporary, accessID);
+           !isAccessType(FieldAccessType::FAT_Literal, accessID) &&
+           !isAccessType(FieldAccessType::FAT_GlobalVariable, accessID);
+  }
+  // not all the accessIDs are registered
+  return (fieldAccessMetadata_.accessIDType_.count(accessID) &&
+          fieldAccessMetadata_.accessIDType_.at(accessID) == fType);
+}
+
+void StencilMetaInformation::moveRegisteredFieldTo(FieldAccessType type, int accessID) {
```
Comment (twicki) [inline src/dawn/IIR/StencilMetaInformation.cpp:261]: can we get a quick line here on what the conventions are?
- Are all the sets disjonit or are there ID's in multiple sets?
- Is every filed required to be (at least) one type?

File: src/dawn/IIR/StencilMetaInformation.cpp [hunk 5]
```diff
@@ -118,35 +101,232 @@ bool StencilMetaInformation::isAccessType(FieldAccessType fType, const std::stri
...
+  // we can not move it into an API field, since the original order would not be preserved
+  DAWN_ASSERT(type != FieldAccessType::FAT_APIField);
+  DAWN_ASSERT_MSG(isFieldType(type), "non field access type can not be moved");
+
+  fieldAccessMetadata_.accessIDType_[accessID] = type;
+
+  if(fieldAccessMetadata_.TemporaryFieldAccessIDSet_.count(accessID)) {
```
Comment (twicki) [inline src/dawn/IIR/StencilMetaInformation.cpp:268]: why does this always happen?

File: src/dawn/IIR/StencilMetaInformation.cpp [hunk 6]
```diff
@@ -118,35 +101,232 @@ bool StencilMetaInformation::isAccessType(FieldAccessType fType, const std::stri
...
+    fieldAccessMetadata_.LiteralAccessIDToNameMap_.emplace(AccessID, name);
+  }
+}
+
+int StencilMetaInformation::insertStmt(bool keepVarNames,
+                                       const std::shared_ptr<VarDeclStmt>& stmt) {
+  int accessID = UIDGenerator::getInstance()->get();
+
+  std::string globalName;
+  if(keepVarNames)
```
Comment (twicki) [inline src/dawn/IIR/StencilMetaInformation.cpp]: can we have if-else `{`s

File: src/dawn/IIR/StencilMetaInformation.cpp [hunk 7]
```diff
@@ -220,31 +405,48 @@ StencilMetaInformation::getStencilFunctionInstantiation(
   return it->second;
 }
 
-// TODO set or emplace ? have a convention
+// TODO private ?
```
Comment (twicki) [inline src/dawn/IIR/StencilMetaInformation.cpp]: needs to be public, right?

File: src/dawn/IIR/StencilMetaInformation.cpp [hunk 8]
```diff
@@ -254,6 +456,21 @@ void StencilMetaInformation::removeAccessID(int AccessID) {
   }
 }
 
+StencilMetaInformation::StencilMetaInformation(const sir::GlobalVariableMap& globalVariables) {
```
Comment (twicki) [inline src/dawn/IIR/StencilMetaInformation.cpp:459]: should there not also be a default ctor? The global variable map is not always required right?

File: src/dawn/IIR/StencilMetaInformation.h [hunk 9]
```diff
@@ -17,9 +17,11 @@
 
 #include "dawn/IIR/Extents.h"
 #include "dawn/IIR/FieldAccessMetadata.h"
+#include "dawn/IIR/InstantiationHelper.h"
```
Comment (twicki) [inline src/dawn/IIR/StencilMetaInformation.h]: move this to the cpp file

File: src/dawn/Optimizer/OptimizerContext.cpp [hunk 10]
```diff
@@ -89,11 +89,12 @@ class StencilDescStatementMapper : public ASTVisitor {
                                         instantiation_->getIIR()->getControlFlowDescriptor()));
     scope_.top()->LocalFieldnameToAccessIDMap = metadata_.getNameToAccessIDMap();
 
+    // TODO redo this, we dont need to copy again the variable map into scope::top
```
Comment (twicki) [inline src/dawn/Optimizer/OptimizerContext.cpp]: Open TODO?

File: src/dawn/Optimizer/PassSetCaches.cpp [hunk 11]
```diff
@@ -249,11 +249,6 @@ bool PassSetCaches::run(const std::shared_ptr<iir::StencilInstantiation>& instan
                                            field.getAccessID()),
               ms);
 
-          DAWN_ASSERT((cacheCandidate.policy_ != iir::Cache::fill &&
```
Comment (twicki) [inline src/dawn/Optimizer/PassSetCaches.cpp:252]: why is this removed?

File: src/dawn/IIR/StencilMetaInformation.cpp [hunk 12]
```diff
@@ -118,35 +101,232 @@ bool StencilMetaInformation::isAccessType(FieldAccessType fType, const std::stri
...
+  int accessID = UIDGenerator::getInstance()->get();
+
+  std::string globalName;
+  if(keepVarNames)
+    globalName = stmt->getName();
+  else
+    globalName = InstantiationHelper::makeLocalVariablename(stmt->getName(), accessID);
+
+  setAccessIDNamePair(accessID, globalName);
+  StmtIDToAccessIDMap_.emplace(stmt->getID(), accessID);
```
Comment (twicki) [inline src/dawn/IIR/StencilMetaInformation.cpp:322]:
```suggestion
  DAWN_ASSERT(!StmtIDToAccessIDMap_.count(stmt->getID()));
  StmtIDToAccessIDMap_.emplace(stmt->getID(), accessID);
```

File: src/dawn/IIR/StencilMetaInformation.cpp [hunk 13]
```diff
@@ -156,16 +336,28 @@ Array3i StencilMetaInformation::getFieldDimensionsMask(int FieldID) const {
...
 }
 
-void StencilMetaInformation::mapStmtToAccessID(const std::shared_ptr<Stmt>& stmt, int accessID) {
+void StencilMetaInformation::eraseStmtToAccessID(std::shared_ptr<Stmt> stmt) {
+  DAWN_ASSERT(StmtIDToAccessIDMap_.count(stmt->getID()));
+  StmtIDToAccessIDMap_.erase(stmt->getID());
+}
+
+void StencilMetaInformation::insertStmtToAccessID(const std::shared_ptr<Stmt>& stmt, int accessID) {
   StmtIDToAccessIDMap_.emplace(stmt->getID(), accessID);
```
Comment (twicki) [inline src/dawn/IIR/StencilMetaInformation.cpp:362]:
```suggestion
DAWN_ASSERT(!StmtIDToAccessIDMap_.count(stmt->getID()));
StmtIDToAccessIDMap_.emplace(stmt->getID(), accessID);
```

**Review by twicki — COMMENTED** [review]
There are still some open points from the last review:
- recovery of codegenproperties.cpp's assert is still missing
- indentation of IIR/CmakeLists still bad (see with any non-qt-editor)
- ControlFlowDescriptor still needs forwarding for insertStmt to remove code-duplication
- DependencyGraphAccess still has an open TODO
- IIR.cpp still has shallow copy in clone
- Question on the open Todo in StencilMetaInformation still open:
    at() will throw OOB if the element is not there. If the element is there, there is alway a 1-to-1 mapping if the insertion method is correct (i.e it checks that we're not overriding
- Emplaces in StencilMetaInformation are still w/o Assertions if they already exist (see review here)
- DoubleSidedMap.h: Do we want two different inserts where one replaces and the other one asserts if the element already exists. This would make it more explicit what we do

File: src/dawn/IIR/StencilInstantiation.cpp (hunk 2 above)
Comment (cosunae) [reply src/dawn/IIR/StencilInstantiation.cpp:51]: it is store only in IIR, but we need it in order to construct certain metadata of the StencilMetaInformation. Therefore it is passed to the ctr that will build that metadata (instead of constructing empty containers and initializing later, which leaves the object in inconsistent state after ctr) but not stored.

File: src/dawn/IIR/StencilInstantiation.h (hunk 3 above)
Comment (cosunae) [reply src/dawn/IIR/StencilInstantiation.h]: done

File: src/dawn/IIR/StencilMetaInformation.cpp (hunk 4 above)
Comment (cosunae) [reply src/dawn/IIR/StencilMetaInformation.cpp:261]: I tried to document the convention in the header of the class that holds the containers

https://github.com/cosunae/dawn/pull/17/files#diff-d1c017d34c8141a817d1aac58ffffa75R155

File: src/dawn/IIR/StencilMetaInformation.cpp (hunk 5 above)
Comment (cosunae) [reply src/dawn/IIR/StencilMetaInformation.cpp:268]: I didnt understand, what is what always happen?

File: src/dawn/IIR/StencilMetaInformation.cpp (hunk 6 above)
Comment (cosunae) [reply src/dawn/IIR/StencilMetaInformation.cpp]: done

File: src/dawn/IIR/StencilMetaInformation.cpp (hunk 7 above)
Comment (cosunae) [reply src/dawn/IIR/StencilMetaInformation.cpp]: yes, removed the TODO

File: src/dawn/IIR/StencilMetaInformation.cpp (hunk 8 above)
Comment (cosunae) [reply src/dawn/IIR/StencilMetaInformation.cpp:459]: I would say yes (the global variables map should always be required), that one should be always required, it can be empty though. Having a default ctr can lead to the global variable related maps if you forget to construct them later on. Building always at ctr time follows the RAII

File: src/dawn/IIR/StencilMetaInformation.h (hunk 9 above)
Comment (cosunae) [reply src/dawn/IIR/StencilMetaInformation.h]: done

File: src/dawn/Optimizer/OptimizerContext.cpp (hunk 10 above)
Comment (cosunae) [reply src/dawn/Optimizer/OptimizerContext.cpp]: ok, I think we should refactor one day this Scope class (and all of the Scope's we have) but I dont dare to touch it for this PR (it is also not related) so I removed the TODO

**Comment by cosunae** [comment]
- [ ]    recovery of codegenproperties.cpp's assert is still missing
I can not recover it. The BC are quite broken, and what happens is that for FW the same BC struct is generated more than one and the this assertion is triggered. This has nothing to do with this PR, so I wouldnt fix BC here. We have to go through it at some point, so that is why I would leave the TODO in the assertion so that we can recover it at some point
- [x]     indentation of IIR/CmakeLists still bad (see with any non-qt-editor)

- [x]    ControlFlowDescriptor still needs forwarding for insertStmt to remove code-duplication
- [x]    DependencyGraphAccess still has an open TODO
- [x]    IIR.cpp still has shallow copy in clone
- [x]    Question on the open Todo in StencilMetaInformation still open:
- [x]    at() will throw OOB if the element is not there. If the element is there, there is alway a 1-to-1 mapping if the insertion method is correct (i.e it checks that we're not overriding

About the assertions below, yes, I agree if we would have a clean state, however I am afraid that when compiling the clang-gridtools many will triggered. I will try in a separate PR, but I dont there to do it here, in case it break
- [ ]    Emplaces in StencilMetaInformation are still w/o Assertions if they already exist (see review here)
- [ ]    DoubleSidedMap.h: Do we want two different inserts where one replaces and the other one asserts if the element already exists. This would make it more explicit what we do

File: src/dawn/Optimizer/PassSetCaches.cpp (hunk 11 above)
Comment (cosunae) [reply src/dawn/Optimizer/PassSetCaches.cpp:252]: there were cases in the clang-gridtools that triggered the assertion and were actually correct. 
Supporting those cases turned into an always true condition. 
If you read the condition after || also you realized that it was ill formed and didnt make sense.

File: src/dawn/IIR/StencilInstantiation.cpp (hunk 2 above)
Comment (twicki) [reply src/dawn/IIR/StencilInstantiation.cpp:51]: Thanks, this is clear now.

File: src/dawn/IIR/StencilMetaInformation.cpp (hunk 4 above)
Comment (twicki) [reply src/dawn/IIR/StencilMetaInformation.cpp:261]: got it

File: src/dawn/IIR/StencilMetaInformation.cpp (hunk 5 above)
Comment (twicki) [reply src/dawn/IIR/StencilMetaInformation.cpp:268]: it's ok. I needed a second look

File: src/dawn/IIR/StencilMetaInformation.cpp (hunk 8 above)
Comment (twicki) [reply src/dawn/IIR/StencilMetaInformation.cpp:459]: makes sense.

File: src/dawn/Optimizer/OptimizerContext.cpp (hunk 10 above)
Comment (twicki) [reply src/dawn/Optimizer/OptimizerContext.cpp]: can we have an issue somewhere so that we don't forget?

**Comment by twicki** [comment]
everything looks good to me. I think we can merge now!

**Comment by twicki** [comment]
we've gone through all of the review process, this can be closed
